File: java/1310_xorQueries.java
class Solution {
    public int[] xorQueries(int[] arr, int[][] queries) {
        int[] record = new int[arr.length+1];
        for(int i=0;i<arr.length;i++)
            record[i+1] = record[i]^arr[i];
        int[] res = new int[queries.length];
        for(int i=0;i<queries.length;i++)
            res[i] = record[queries[i][0]]^record[queries[i][1]+1];
        return res;
    }
}

File: cpp/1310_xorQueries.cpp
// 利用前缀异或的性质（a^a)^b = b，不需要每次重复计算
// arr[left] ^...^ arr[right] = 
// [(arr[0]^...^arr[left-1]) ^ (arr[0]^...^arr[left-1])]^ (arr[left] ^...^ arr[right])=
// (arr[0]^...^arr[left-1]) ^ (arr[0]^...^arr[right])
class Solution {
public:
    vector<int> xorQueries(vector<int>& arr, vector<vector<int>>& queries) {
        vector<int> record(arr.size()+1,0); // 0^num = num
        for(int i=0;i<arr.size();i++)
            record[i+1] = record[i]^arr[i];
        vector<int>res;
        for(int i=0;i<queries.size();i++)
            res.push_back(record[queries[i][0]]^record[queries[i][1]+1]);
        return res;
    }
};
